### PR TITLE
update Token on NewClient

### PR DESCRIPTION
**Pull request by jgsqware**
On NewClient, in case of ExpiredToken, regenerate a new Token and update the current Token

### CODE REVIEW

**Comment by jgsqware** [comment]
AppVeyor apparently has not been relauched after the last commit. The error AppVeyor show is fixed

**Comment by zmb3** [comment]
I'm not sure that this is necessary.  The client we obtain from the oauth2 package will refresh the token automatically.

Per the[ docs](https://godoc.org/golang.org/x/oauth2#Config.Client):
> Client returns an HTTP client using the provided token. The token will auto-refresh as necessary.

**Comment by jgsqware** [comment]
Yep, but the retrieve client don't have the token accessible. I use it to
save to token for further request

Le ven. 3 mars 2017 16:01, Zac Bergquist <notifications@github.com> a
écrit :

**Comment by zmb3** [comment]
I might just not understand your use case.  You shouldn't need to access the token as long as you use the client returned by `NewClient` - everything is handled transparently to you.

What do you need access to the token for?  We might be able to make it easier on you.

**Comment by jgsqware** [comment]
I use the lib in a cli.
The first authorization save the token for later use, then on next run a
reuse that token and NewClient renew it. But I cannot save that renewed
token.

Le ven. 3 mars 2017 16:13, Zac Bergquist <notifications@github.com> a
écrit :

**Comment by zmb3** [comment]
Okay, that makes sense.

What if we added a `func (c *Client) Token() *oauth2.Token` instead?  I feel like thats a more explicit API than modifying the token passed in to `NewClient`.

**Comment by jgsqware** [comment]
Ok I can update my PR if needed

Le ven. 3 mars 2017 16:18, Zac Bergquist <notifications@github.com> a
écrit :

File: examples/authenticate/authenticate.go [hunk 1]
```diff
@@ -60,6 +60,9 @@ func completeAuth(w http.ResponseWriter, r *http.Request) {
 	}
 	// use the token to get an authenticated client
 	client := auth.NewClient(tok)
+	if err != nil {
```
Comment (zmb3) [inline examples/authenticate/authenticate.go]: This needs to be removed.

File: auth.go [hunk 2]
```diff
@@ -151,3 +151,13 @@ func (a Authenticator) NewClient(token *oauth2.Token) Client {
 		http: client,
 	}
 }
+
+// Token return the current client token
+func (c *Client) Token() (*oauth2.Token, error) {
+	t, err := c.http.Transport.(*oauth2.Transport).Source.Token()
```
Comment (zmb3) [inline auth.go]: Let's check the type assertion just to be safe:
```
transport, ok := c.http.Transport.(*ouath2.Transport)
if !ok {
   return nil, errors.New("...")
}
t, err := transport.Source.Token()
// ...
```

File: auth.go [hunk 3]
```diff
@@ -151,3 +151,13 @@ func (a Authenticator) NewClient(token *oauth2.Token) Client {
 		http: client,
 	}
 }
+
+// Token return the current client token
```
Comment (zmb3) [inline auth.go]: How about: 
`// Token gets the client's current token.`

**Comment by chances** [comment]
I have a similar use case: Storing the refresh token in a database for future use.

Anything we can do to help get this PR merged?

**Comment by zmb3** [comment]
Sorry this slipped through the cracks. I never saw the review comments addressed. Looks good!

**Pull request merged**